File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/rest-api/rest-api.js
//
// This is only a SKELETON file for the 'Rest API' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class RestAPI {
  constructor(database = { users: [] }) {
    this.db = database.users ? database : { users: [] };
  }

  get(url) {
    if (url.startsWith('/users')) {
      const match = url.match(/\?users=(.*)/);
      const users = match ? match[1].split(',') : null;
      
      let result = this.db.users;
      if (users) {
        result = result.filter(user => users.includes(user.name));
      }
      return { users: result.sort((a, b) => a.name.localeCompare(b.name)) };
    }
    throw new Error('Invalid URL');
  }

  post(url, payload) {
    if (url === '/add') {
      const newUser = {
        name: payload.user,
        owes: {},
        owed_by: {},
        balance: 0
      };
      
      if (this.db.users.some(user => user.name === payload.user)) {
        throw new Error('User already exists');
      }
      
      this.db.users.push(newUser);
      return newUser;
    }
    
    if (url === '/iou') {
      const { lender, borrower, amount } = payload;
      let lenderUser = this.db.users.find(u => u.name === lender);
      let borrowerUser = this.db.users.find(u => u.name === borrower);
      
      if (!lenderUser || !borrowerUser) {
        throw new Error('User not found');
      }

      // Handle existing debts in opposite direction
      const existingDebt = borrowerUser.owed_by[lender] || 0;
      if (existingDebt > 0) {
        if (existingDebt >= amount) {
          // Reduce existing debt
          borrowerUser.owed_by[lender] -= amount;
          lenderUser.owes[borrower] -= amount;
          if (borrowerUser.owed_by[lender] === 0) {
            delete borrowerUser.owed_by[lender];
            delete lenderUser.owes[borrower];
          }
        } else {
          // Clear existing debt and create new debt in opposite direction
          delete borrowerUser.owed_by[lender];
          delete lenderUser.owes[borrower];
          const remaining = amount - existingDebt;
          borrowerUser.owes[lender] = remaining;
          lenderUser.owed_by[borrower] = remaining;
        }
      } else {
        // Add to existing debt or create new debt
        borrowerUser.owes[lender] = (borrowerUser.owes[lender] || 0) + amount;
        lenderUser.owed_by[borrower] = (lenderUser.owed_by[borrower] || 0) + amount;
      }

      // Update balances
      lenderUser.balance = Object.values(lenderUser.owed_by).reduce((sum, val) => sum + val, 0) -
                          Object.values(lenderUser.owes).reduce((sum, val) => sum + val, 0);
      borrowerUser.balance = Object.values(borrowerUser.owed_by).reduce((sum, val) => sum + val, 0) -
                            Object.values(borrowerUser.owes).reduce((sum, val) => sum + val, 0);

      return {
        users: [lenderUser, borrowerUser].sort((a, b) => a.name.localeCompare(b.name))
      };
    }
    
    throw new Error('Invalid URL');
  }
}

File: src/2025-03-16-08-42-59--a-deepclaude-full-python-0316-1/python/exercises/practice/rest-api/rest_api.py
import json

class RestAPI:
    def __init__(self, database=None):
        self.database = database if database else {"users": []}

    def get(self, url, payload=None):
        if url == "/users":
            if payload:
                data = json.loads(payload)
                requested_users = data["users"]
                users = [user for user in self.database["users"] if user["name"] in requested_users]
            else:
                users = self.database["users"]
            return json.dumps({"users": users})
        return json.dumps({})

    def post(self, url, payload=None):
        if url == "/add":
            data = json.loads(payload)
            new_user = {
                "name": data["user"],
                "owes": {},
                "owed_by": {},
                "balance": 0.0
            }
            self.database["users"].append(new_user)
            return json.dumps(new_user)
        
        elif url == "/iou":
            data = json.loads(payload)
            lender_name = data["lender"]
            borrower_name = data["borrower"]
            amount = data["amount"]
            
            # Find users
            lender = next(user for user in self.database["users"] if user["name"] == lender_name)
            borrower = next(user for user in self.database["users"] if user["name"] == borrower_name)
            
            # Check if borrower already owes lender
            if lender_name in borrower["owes"]:
                borrower_owes = borrower["owes"][lender_name]
                if amount <= borrower_owes:
                    # Reduce existing debt
                    borrower["owes"][lender_name] -= amount
                    lender["owed_by"][borrower_name] -= amount
                    
                    # Remove if zero
                    if borrower["owes"][lender_name] == 0:
                        del borrower["owes"][lender_name]
                        del lender["owed_by"][borrower_name]
                else:
                    # Debt reversal
                    del borrower["owes"][lender_name]
                    del lender["owed_by"][borrower_name]
                    
                    difference = amount - borrower_owes
                    lender["owes"][borrower_name] = difference
                    borrower["owed_by"][lender_name] = difference
            # Check if lender already owes borrower
            elif borrower_name in lender["owes"]:
                lender_owes = lender["owes"][borrower_name]
                lender["owes"][borrower_name] += amount
                borrower["owed_by"][lender_name] += amount
            else:
                # New debt
                if borrower_name not in lender["owed_by"]:
                    lender["owed_by"][borrower_name] = 0
                if lender_name not in borrower["owes"]:
                    borrower["owes"][lender_name] = 0
                
                lender["owed_by"][borrower_name] += amount
                borrower["owes"][lender_name] += amount
            
            # Update balances
            lender["balance"] = sum(lender["owed_by"].values()) - sum(lender["owes"].values())
            borrower["balance"] = sum(borrower["owed_by"].values()) - sum(borrower["owes"].values())
            
            # Clean up empty dicts
            for user in [lender, borrower]:
                if not user["owes"]:
                    user["owes"] = {}
                if not user["owed_by"]:
                    user["owed_by"] = {}
            
            return json.dumps({"users": sorted([lender, borrower], key=lambda x: x["name"])})
